### PR TITLE
attempt to add python function decoder

### PR DESCRIPTION
**Pull request by koubaa**
Adds a python action decoder to your branch
Theres' a lot more work here to do to really make this fully featured but I wanted to submit the PR here while your branch hasn't yet been merged into pythonnet/pythonnet for feedback.

### CODE REVIEW

File: src/runtime/Codecs/FunctionCodec.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    //converts python functions to C# actions
+    class FunctionCodec : IPyObjectDecoder
+    {
+        public static FunctionCodec Instance { get; } = new FunctionCodec();
+        public bool CanDecode(PyObject objectType, Type targetType)
+        {
+            if (!objectType.IsCallable()) return false;
+
+            //TODO - handle nonzero arguments.
+            var args = targetType.GetGenericArguments();
```
Comment (lostmsu) [inline src/runtime/Codecs/FunctionCodec.cs]: Before trying anything else, you should check if your codec actually supports `targetType` too.

File: src/runtime/Codecs/FunctionCodec.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            return args.Length == 0;
+        }
+
+        public bool TryDecode<T>(PyObject pyObj, out T value)
+        {
+            Action action = () =>
```
Comment (lostmsu) [inline src/runtime/Codecs/FunctionCodec.cs]: Here I'd also check if `T` is `Action` before doing anything else.

**Review by lostmsu — CHANGES_REQUESTED** [review]
Of course this needs generics handling to be accepted here or in upstream (or rather any delegate type handling). But looks good for singular use.

**Comment by koubaa** [comment]
@lostmsu generics are hard, especially since I don't have much experience with them in C# (I'm much more familiar with c++).  Anyways I am trying out your suggestions.  Right now the hardest part is determining how many arguments for System.Action to use in TryDecode.  I think each argument has to be "object" because python functions don't care about the type.  I'm ignoring **kwargs for now.

Anyways I pushed and would appreciate a review if you are willing.  Or if you want you can tell me to bug off and read the manual on C# generics, I really should..

**Comment by lostmsu** [comment]
@koubaa you might want to check this out: https://stackoverflow.com/questions/429552/can-i-get-the-signature-of-a-c-sharp-delegate-by-its-type

**Comment by lostmsu** [comment]
Closing this, as Codecs are now in upstream.

**Pull request closed without merging**